File: C++/Parcial.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <queue>
#include <stack>

using namespace std;

struct Libro {
    string titulo;
    string autor;
    int anoEdicion;
    string editorial;
    string isbn;
    int paginas;
    Libro* siguiente;
};

struct Lector {
    string nombre;
    int dni;
    string libroSolicitado; 
    string autor;
    int anoEdicion; 
    string editorial; 
    string isbn; 
    int paginas; 
    Lector* siguiente;
};

// Lista de libros (Lista enlazada simple)
Libro* listaLibros = nullptr;

// Cola de lectores que solicitaron libros
queue<Lector*> colaSolicitudes;

// Pila de historial de operaciones
stack<string> historial;

// Función para agregar un libro a la lista enlazada
void agregarLibro(Libro* &lista, const string& titulo, const string& autor, int ano, const string& editorial, const string& isbn, int paginas) {
    Libro* nuevoLibro = new Libro{titulo, autor, ano, editorial, isbn, paginas, nullptr};
    if (!lista) {
        lista = nuevoLibro;
    } else {
        Libro* actual = lista;
        while (actual->siguiente) {
            actual = actual->siguiente;
        }
        actual->siguiente = nuevoLibro;
    }
    string movimiento = "Libro agregado: " + titulo;
    historial.push(movimiento);
    cout << "Movimiento: " << movimiento << endl; // Mostrar movimiento
    // Guardar movimiento en el archivo de historial
    ofstream archivoHistorial("historial.txt", ios::app);
    archivoHistorial << movimiento << endl;
    archivoHistorial.close();
}

// Función para ordenar libros por título (Método de burbuja)
void ordenarLibrosPorTitulo(Libro* &lista) {
    if (!lista || !lista->siguiente) return;
    bool ordenado = false;
    while (!ordenado) {
        ordenado = true;
        Libro* actual = lista;
        while (actual->siguiente) {
            if (actual->titulo > actual->siguiente->titulo) {
                swap(actual->titulo, actual->siguiente->titulo);
                swap(actual->autor, actual->siguiente->autor);
                swap(actual->anoEdicion, actual->siguiente->anoEdicion);
                swap(actual->editorial, actual->siguiente->editorial);
                swap(actual->isbn, actual->siguiente->isbn);
                swap(actual->paginas, actual->siguiente->paginas);
                ordenado = false;
            }
            actual = actual->siguiente;
        }
    }
    string movimiento = "Libros ordenados";
    historial.push(movimiento);
    cout << "Movimiento: " << movimiento << endl; // Mostrar movimiento
    // Guardar movimiento en el archivo de historial
    ofstream archivoHistorial("historial.txt", ios::app);
    archivoHistorial << movimiento << endl;
    archivoHistorial.close();
}

// Función para buscar un libro por título, autor o ISBN
Libro* buscarLibro(Libro* lista, const string& clave, const string& tipo) {
    Libro* actual = lista;
    while (actual) {
        if ((tipo == "titulo" && actual->titulo == clave) || 
            (tipo == "autor" && actual->autor == clave) ||
            (tipo == "isbn" && actual->isbn == clave)) {
            return actual;
        }
        actual = actual->siguiente;
    }
    return nullptr;
}

// Función para mostrar la información de un libro
void mostrarLibro(Libro* libro) {
    if (libro) {
        cout << "\n--- Información del Libro ---" << endl;
        cout << "Título: " << libro->titulo << endl;
        cout << "Autor: " << libro->autor << endl;
        cout << "Año de Edición: " << libro->anoEdicion << endl;
        cout << "Editorial: " << libro->editorial << endl;
        cout << "ISBN: " << libro->isbn << endl;
        cout << "Páginas: " << libro->paginas << endl;
    } else {
        cout << "Libro no encontrado." << endl;
    }
}

// Función para realizar la búsqueda por título, autor o ISBN
void buscarLibroPorClave() {
    string clave, tipo;
    int opcionBusqueda;
    
    cout << "\n--- Buscar Libro ---" << endl;
    cout << "1. Buscar por Título" << endl;
    cout << "2. Buscar por Autor" << endl;
    cout << "3. Buscar por ISBN" << endl;
    cout << "Ingrese una opción: ";
    cin >> opcionBusqueda;
    cin.ignore();

    if (opcionBusqueda == 1) {
        tipo = "titulo";
        cout << "Ingrese el título del libro: ";
    } else if (opcionBusqueda == 2) {
        tipo = "autor";
        cout << "Ingrese el autor del libro: ";
    } else if (opcionBusqueda == 3) {
        tipo = "isbn";
        cout << "Ingrese el ISBN del libro: ";
    } else {
        cout << "Opción no válida." << endl;
        return;
    }
    getline(cin, clave);

    Libro* libro = buscarLibro(listaLibros, clave, tipo);
    mostrarLibro(libro);
}

// Solicita un libro (Encola el lector en espera y elimina el libro de la lista)
void solicitarLibro(queue<Lector*> &cola, const string& nombre, int dni, const string& libroSolicitado) {
    Lector* nuevoLector = new Lector;
    Libro* libro = buscarLibro(listaLibros, libroSolicitado, "titulo");
    
    if (libro) {
        nuevoLector->nombre = nombre;
        nuevoLector->dni = dni;
        nuevoLector->libroSolicitado = libro->titulo;
        nuevoLector->autor = libro->autor;
        nuevoLector->anoEdicion = libro->anoEdicion;
        nuevoLector->editorial = libro->editorial;
        nuevoLector->isbn = libro->isbn;
        nuevoLector->paginas = libro->paginas;
        nuevoLector->siguiente = nullptr;

        cola.push(nuevoLector);

        // Eliminar el libro de la lista
        Libro* anterior = nullptr;
        Libro* actual = listaLibros;

        while (actual) {
            if (actual->titulo == libroSolicitado) {
                if (anterior) {
                    anterior->siguiente = actual->siguiente;
                } else {
                    listaLibros = actual->siguiente; 
                }
                delete actual; // Eliminar el libro de la lista
                break;
            }
            anterior = actual;
            actual = actual->siguiente;
        }

        string movimiento = "Libro solicitado: " + libroSolicitado;
        historial.push(movimiento);
        cout << "Movimiento: " << movimiento << endl; // Mostrar movimiento
        // Guardar movimiento en el archivo de historial
        ofstream archivoHistorial("historial.txt", ios::app);
        archivoHistorial << movimiento << endl;
        archivoHistorial.close();
    } else {
        cout << "El libro solicitado no está disponible." << endl;
        delete nuevoLector; // Liberar memoria si no se solicitó
    }
}

// Muestra los libros que han sido solicitados
void mostrarLibrosSolicitados(queue<Lector*> cola) {
    cout << "\n--- Libros Solicitados ---" << endl;
    int i = 1;
    while (!cola.empty()) {
        Lector* lector = cola.front();
        cout << i++ << ". " << lector->libroSolicitado << " (Autor: " << lector->autor << ")" << endl;
        cola.pop();
    }
}

// Devuelve un libro específico de la cola
void devolverLibro(queue<Lector*> &cola) {
    if (!cola.empty()) {
        mostrarLibrosSolicitados(cola);
        
        int seleccion;
        cout << "Seleccione el número del libro que desea devolver: ";
        cin >> seleccion;

        // Validar selección
        if (seleccion < 1 || seleccion > cola.size()) {
            cout << "Selección no válida." << endl;
            return;
        }

        // Eliminar el lector de la cola
        Lector* lectorAtendido = nullptr;
        for (int i = 1; i <= seleccion; ++i) {
            lectorAtendido = cola.front();
            if (i == seleccion) break;
            cola.pop();
        }

        // Eliminar el lector de la cola (devolver el libro)
        cola.pop();

        // Agregar el libro devuelto a la lista de libros con los datos originales
        agregarLibro(listaLibros, lectorAtendido->libroSolicitado, lectorAtendido->autor, 
                     lectorAtendido->anoEdicion, lectorAtendido->editorial, 
                     lectorAtendido->isbn, lectorAtendido->paginas);
        
        string movimiento = "Libro devuelto: " + lectorAtendido->libroSolicitado;"por " + lectorAtendido->nombre;
        historial.push(movimiento);
        cout << "Movimiento: " << movimiento << endl;
        
        ofstream archivoHistorial("historial.txt", ios::app);
        archivoHistorial << movimiento << endl;
        archivoHistorial.close();

        delete lectorAtendido; // Liberar memoria del lector atendido
    } else {
        cout << "No hay libros solicitados para devolver." << endl;
    }
}

// Guarda los libros actuales en el archivo biblioteca.txt
void guardarLibros(Libro* lista) {
    ofstream archivo("biblioteca.txt");
    if (!archivo) {
        cerr << "No se pudo abrir el archivo para guardar." << endl;
        return;
    }

    Libro* actual = lista;
    while (actual) {
        archivo << actual->titulo << "," << actual->autor << "," << actual->anoEdicion << ","
                << actual->editorial << "," << actual->isbn << "," << actual->paginas << endl;
        actual = actual->siguiente;
    }
    archivo.close();
    cout << "Libros guardados correctamente." << endl;
}

// Muestra el historial de movimientos
void mostrarHistorial() {
    cout << "\n--- Historial de Movimientos ---" << endl;
    stack<string> temp = historial;
    while (!temp.empty()) {
        cout << temp.top() << endl;
        temp.pop();
    }
}

// Vacía la biblioteca liberando la memoria
void vaciarBiblioteca(Libro* &lista) {
    while (lista) {
        Libro* temp = lista;
        lista = lista->siguiente;
        delete temp;
    }
    cout << "Biblioteca vaciada." << endl;
}

int main() {
    int opcion;

    do {
        cout << "\n--- Gestión de Biblioteca ---" << endl;
        cout << "1. Agregar libro" << endl;
        cout << "2. Solicitar libro" << endl;
        cout << "3. Devolver libro" << endl;
        cout << "4. Ordenar libros por título" << endl;
        cout << "5. Buscar libro" << endl;
        cout << "6. Guardar libros" << endl;
        cout << "7. Mostrar historial" << endl;
        cout << "8. Vaciar biblioteca" << endl;
        cout << "0. Salir" << endl;
        cout << "Ingrese una opción: "; cin >> opcion; cin.ignore();

        if (opcion == 1) {
            string titulo, autor, editorial, isbn;
            int ano, paginas;
            cout << "Título: "; getline(cin, titulo);
            cout << "Autor: "; getline(cin, autor);
            cout << "Año de edición: "; cin >> ano;
            cin.ignore();
            cout << "Editorial: "; getline(cin, editorial);
            cout << "ISBN: "; getline(cin, isbn);
            cout << "Número de páginas: "; cin >> paginas; cin.ignore();
            agregarLibro(listaLibros, titulo, autor, ano, editorial, isbn, paginas);
        } else if (opcion == 2) {
            string nombreLibro, nombreLector;
            int dniLector;
            cout << "Ingrese su nombre: "; getline(cin, nombreLector);
            cout << "Ingrese su DNI: "; cin >> dniLector; cin.ignore();
            cout << "Ingrese el título del libro solicitado: "; getline(cin, nombreLibro);
            solicitarLibro(colaSolicitudes, nombreLector, dniLector, nombreLibro);
        } else if (opcion == 3) {
            devolverLibro(colaSolicitudes);
        } else if (opcion == 4) {
            ordenarLibrosPorTitulo(listaLibros);
        } else if (opcion == 5) {
            buscarLibroPorClave();
        } else if (opcion == 6) {
            guardarLibros(listaLibros);
        } else if (opcion == 7) {
            mostrarHistorial();
        } else if (opcion == 8) {
            vaciarBiblioteca(listaLibros);
        }
    } while (opcion != 0);

    // Liberar memoria antes de salir
    while (listaLibros) {
        Libro* temp = listaLibros;
        listaLibros = listaLibros->siguiente;
        delete temp;
    }

    while (!colaSolicitudes.empty()) {
        delete colaSolicitudes.front(); // Liberar memoria del lector que solicito el libro
        colaSolicitudes.pop();
    }

    return 0;
}

File: Python/Parcial.py
class Libro:
    def __init__(self, titulo, autor, ano, editorial, isbn, paginas):
        self.titulo = titulo
        self.autor = autor
        self.ano_edicion = ano
        self.editorial = editorial
        self.isbn = isbn
        self.paginas = paginas

    def __str__(self):
        return f"Título: {self.titulo}, Autor: {self.autor}, Año: {self.ano_edicion}, Editorial: {self.editorial}, ISBN: {self.isbn}, Páginas: {self.paginas}"


class Lector:
    def __init__(self, nombre, dni, libro_solicitado):
        self.nombre = nombre
        self.dni = dni
        self.libro_solicitado = libro_solicitado


class Biblioteca:
    def __init__(self):
        self.libros = self.cargar_libros()
        self.solicitudes = []
        self.historial = []

    def cargar_libros(self):
        libros = []
        try:
            with open("libros_agregados.txt", "r") as f:
                for line in f:
                    datos = line.strip().split(',')
                    if len(datos) == 6:
                        libro = Libro(datos[0], datos[1], int(datos[2]), datos[3], datos[4], int(datos[5]))
                        libros.append(libro)
        except FileNotFoundError:
            print("El archivo de libros no se encontró.")
        return libros

    def agregar_libro(self, libro):
        self.libros.append(libro)
        movimiento = f"Libro agregado: {libro.titulo}"
        self.historial.append(movimiento)
        print(f"Movimiento: {movimiento}")
        self.guardar_libros()  # Guardar en archivo después de agregar
        self.guardar_historial()

    def guardar_historial(self):
        with open("historial_movimientos.txt", "a") as f:
            for movimiento in self.historial:
                f.write(movimiento + "\n")
        self.historial = []  # Limpiar el historial después de guardarlo

    def solicitar_libro(self, nombre, dni, titulo):
        libro = self.buscar_libro(titulo)
        if libro:
            lector = Lector(nombre, dni, libro)
            self.solicitudes.append(lector)
            self.libros.remove(libro)  # Eliminar libro de la lista
            movimiento = f"Libro solicitado: {titulo} por {nombre}"
            self.historial.append(movimiento)
            print(f"Movimiento: {movimiento}")
            self.guardar_historial()
            self.guardar_libros()  # Guardar cambios en el archivo de libros
        else:
            print("El libro solicitado no está disponible.")

    def devolver_libro(self):
        if self.solicitudes:
            self.mostrar_libros_solicitados()
            seleccion = int(input("Seleccione el número del libro que desea devolver: ")) - 1

            if 0 <= seleccion < len(self.solicitudes):
                lector_atendido = self.solicitudes[seleccion]
                self.solicitudes.remove(lector_atendido)
                self.agregar_libro(lector_atendido.libro_solicitado)  # Reagregar el libro a la biblioteca
                movimiento = f"Libro devuelto: {lector_atendido.libro_solicitado.titulo} por {lector_atendido.nombre}"
                self.historial.append(movimiento)
                print(f"Movimiento: {movimiento}")
                self.guardar_historial()
                self.guardar_libros()  # Guardar cambios en el archivo de libros
            else:
                print("Selección no válida.")
        else:
            print("No hay libros solicitados para devolver.")

    def buscar_libro(self, clave):
        for libro in self.libros:
            if libro.titulo.lower() == clave.lower():
                return libro
        return None

    def mostrar_libros_solicitados(self):
        print("\n--- Libros Solicitados ---")
        for i, lector in enumerate(self.solicitudes, start=1):
            print(f"{i}. {lector.libro_solicitado.titulo} (Autor: {lector.libro_solicitado.autor})")

    def guardar_libros(self):
        with open("libros_agregados.txt", "w") as f:
            for libro in self.libros:
                f.write(f"{libro.titulo},{libro.autor},{libro.ano_edicion},{libro.editorial},{libro.isbn},{libro.paginas}\n")
        print("Libros guardados correctamente.")

    def mostrar_historial(self):
        print("\n--- Historial de Movimientos ---")
        try:
            with open("historial_movimientos.txt", "r") as f:
                for line in f:
                    print(line.strip())
        except FileNotFoundError:
            print("No hay historial disponible.")

    def vaciar_historial(self):
        with open("historial_movimientos.txt", "w") as f:
            f.truncate(0)  # Vaciar el contenido del archivo
        print("Historial vaciado correctamente.")

    def vaciar_biblioteca(self):
        self.libros.clear()
        self.guardar_libros()  # Guardar cambios en el archivo de libros
        print("Biblioteca vaciada.")

    def ordenar_libros(self):
        self.libros.sort(key=lambda x: x.titulo)
        print("Libros ordenados por título.")

    def mostrar_libros(self):
        print("\n--- Libros en la Biblioteca ---")
        for libro in self.libros:
            print(libro)


def main():
    biblioteca = Biblioteca()

    while True:
        print("\n--- Gestión de Biblioteca ---")
        print("1. Agregar libro")
        print("2. Solicitar libro")
        print("3. Devolver libro")
        print("4. Ordenar libros por título")
        print("5. Buscar libro")
        print("6. Guardar libros")
        print("7. Mostrar historial")
        print("8. Vaciar historial")  # Opción para vaciar historial
        print("9. Vaciar biblioteca")
        print("0. Salir")
        opcion = input("Ingrese una opción: ")

        if opcion == "1":
            titulo = input("Título: ")
            autor = input("Autor: ")
            ano = int(input("Año de edición: "))
            editorial = input("Editorial: ")
            isbn = input("ISBN: ")
            paginas = int(input("Número de páginas: "))
            libro = Libro(titulo, autor, ano, editorial, isbn, paginas)
            biblioteca.agregar_libro(libro)
        elif opcion == "2":
            nombre = input("Ingrese su nombre: ")
            dni = int(input("Ingrese su DNI: "))
            titulo = input("Ingrese el título del libro solicitado: ")
            biblioteca.solicitar_libro(nombre, dni, titulo)
        elif opcion == "3":
            biblioteca.devolver_libro()
        elif opcion == "4":
            biblioteca.ordenar_libros()
        elif opcion == "5":
            titulo = input("Ingrese el título del libro a buscar: ")
            libro = biblioteca.buscar_libro(titulo)
            if libro:
                print(f"Libro encontrado: {libro}")
            else:
                print("Libro no encontrado.")
        elif opcion == "6":
            biblioteca.guardar_libros()
        elif opcion == "7":
            biblioteca.mostrar_historial()
        elif opcion == "8":  # Vaciar historial
            biblioteca.vaciar_historial()
        elif opcion == "9":
            biblioteca.vaciar_biblioteca()
        elif opcion == "0":
            break
        else:
            print("Opción no válida.")

if __name__ == "__main__":
    main()
